browser/src/libs/code_intelligence/code_intelligence.tsx

```diff
@@ -942,46 +943,34 @@ export function handleCodeHost({
-                const update = (decorations?: TextDocumentDecoration[] | null): void => {
-                    try {
-                        decorationsByLine = applyDecorations(
-                            domFunctions,
-                            element,
-                            decorations || [],
-                            decorationsByLine,
-                            'base'
-                        )
-                    } catch (err) {
-                        console.error('Could not apply base decorations to code view', codeViewEvent.element, err)
-                    }
+
+            // Apply decorations coming from extensions
+            if (!minimalUI) {
+                if (diffOrBlobInfo.type === 'blob') {
+                    applyDecorationsForFileInfo(diffOrBlobInfo)
+                } else if (diffHasHead(diffOrBlobInfo)) {
+                    applyDecorationsForFileInfo(diffOrBlobInfo.head, 'head')
+                } else if (diffHasBase(diffOrBlobInfo)) {
+                    applyDecorationsForFileInfo(diffOrBlobInfo.base, 'base')
                 }
-                codeViewEvent.subscriptions.add(
-                    extensionsController.services.textDocumentDecoration
-                        .getDecorations(
-                            toTextDocumentIdentifier({
-                                repoName: fileInfo.baseRepoName || fileInfo.repoName, // not sure if all code hosts set baseRepoName
-                                commitID: fileInfo.baseCommitID,
-                                filePath: fileInfo.baseFilePath,
-                            })
-                        )
-                        // Make sure decorations get cleaned up on unsubscription
-                        .pipe(finalize(update))
-                        // The nested subscribe cannot be replaced with a switchMap()
-                        // We manage the subscription correctly.
-                        // eslint-disable-next-line rxjs/no-nested-subscribe
-                        .subscribe(update)
-                )
             }
 
             // Add hover code intelligence
-            const resolveContext: ContextResolver<RepoSpec & RevSpec & FileSpec & ResolvedRevSpec> = ({ part }) => ({
-                repoName: part === 'base' ? fileInfo.baseRepoName || fileInfo.repoName : fileInfo.repoName,
-                commitID: part === 'base' ? fileInfo.baseCommitID! : fileInfo.commitID,
-                filePath: part === 'base' ? fileInfo.baseFilePath || fileInfo.filePath : fileInfo.filePath,
-                rev: part === 'base' ? fileInfo.baseRev || fileInfo.baseCommitID! : fileInfo.rev || fileInfo.commitID,
-            })
+            const resolveContext: ContextResolver<RepoSpec & RevSpec & FileSpec & ResolvedRevSpec> = ({ part }) => {
+                if (diffOrBlobInfo.type === 'blob') {
+                    return ensureRev(diffOrBlobInfo)
+                } else if (diffHasHead(diffOrBlobInfo) && part === 'head') {
```

Comment:
This should be possible, as long as `head` is defined in all members of the union type (just as explicit `head: undefined` in some)
```suggestion
                } else if (diffOrBlobInfo.head && part === 'head') {
```

Reply:
Alternatively the type guard could be inlined (since I'm getting rid of the tagged union anyway) as `'head' in diffOrBlobInfo` here, right?

Reply:
Not quite, see my other comment: https://github.com/sourcegraph/sourcegraph/pull/10779#discussion_r430041413